### PR TITLE
[stdune] Set.to_list_map_rev

### PR DESCRIPTION
**Pull request by rgrinberg**
To complement of_list_map

### CODE REVIEW

**Comment by aalekseyev** [comment]
Do all those lists actually end up reversed as a result of this PR? Can we avoid it? I find having to work with reversed lists and remembering which list is reversed quite error-prone, so I'd much prefer if we just had an extra List.rev for now.
In the ideal world there would be a `Map.fold_right`, but we don't have that right now, it seems. :-(

**Comment by rgrinberg** [comment]
> Do all those lists actually end up reversed as a result of this PR? Can we avoid it?

The extra intermediate list would defeat the point of this optimization.

> I find having to work with reversed lists and remembering which list is reversed quite error-prone, so I'd much prefer if we just had an extra List.rev for now.

That's why I didn't change all instances of `to_list  . List.map`. Only the ones were it's fairly obvious the order doesn't matter. I could reduce usage further and stick to only using it in hot spots.

**Comment by aalekseyev** [comment]
> The extra intermediate list would defeat the point of this optimization.

I understand we currently have three intermediate lists:
- the one returned by to_list
- the one constructed by map_rev
- the one constructed by rev

You'd still be removing two of them, and it's clearly possible to remove the third in the future without changing the semantics.

> That's why I didn't change all instances ...

I'd much rather change all instances and slash allocation by two thirds without having to think about reversed lists.

**Comment by rgrinberg** [comment]
Alright, I added the extra `List.rev`

**Comment by aalekseyev** [comment]
> without changing the semantics

I guess that's not quite right because `fold_right` changes the traversal order.

Still, I don't like those "rev"s. I think they make the code confusing and we should reserve them for places that are performance hotspots.

File: src/dune_engine/dep.ml [hunk 1]
```diff
@@ -63,9 +63,7 @@ module T = struct
     | Alias a -> Some (File (trace_file (Path.build (Alias.stamp_file a))))
     | File_selector dir_glob ->
       let id = File_selector.to_dyn dir_glob
-      and files =
-        eval_pred dir_glob |> Path.Set.to_list |> List.map ~f:trace_file
-      in
+      and files = eval_pred dir_glob |> Path.Set.to_list_map ~f:trace_file in
```
Comment (rgrinberg) [inline src/dune_engine/dep.ml:66]: @aalekseyev this bit is a hot spot. Shall we just write an explicit fold here perhaps?

**Comment by nojb** [comment]
Apologies for the drive-by commenting, but in the past, I found that the only way to reason easily about all this "rev" stuff is to have a dedicated type `'a RevList.t` for "reversed lists".

**Review by aalekseyev — APPROVED** [review]

File: src/dune_engine/dep.ml (hunk 1 above)
Comment (aalekseyev) [reply src/dune_engine/dep.ml:66]: I don't mind the existence of `to_list_map_rev` if that's your preferred solution, I just think it should be used sparingly. 
Seems fine to use it here.
A fold written inline seems fine too.

**Comment by rgrinberg** [comment]
> Apologies for the drive-by commenting, but in the past, I found that the only way to reason easily about all this "rev" stuff is to have a dedicated type 'a RevList.t for "reversed lists".

Yeah, i've had something like that in mind as well:
```ocaml
type 'a revlist =
  | In_order of 'a list
  | Rev of 'a list

val to_list : 'a revlist -> 'a list (* reverses if necessary *)
val rev : 'a revlist -> 'a revlist (* flips in_order and rev *)
val map : 'a revlist -> f:('a -> 'b) -> 'a revlist
..
```

But it seems more trouble than its worth.

**Comment by aalekseyev** [comment]
We have a much more lightweight version here: https://github.com/janestreet/reversed_list/blob/master/src/reversed_list.mli

```
type 'a t =
  | []
  | ( :: ) of 'a * 'a t

val rev : 'a t -> 'a list
val rev_append : 'a t -> 'a list -> 'a list
```

I think it's useful for cases where you actually do care about the order, to help make sure that it's correct.

It's not so useful when you want to show that the order doesn't matter, though, for example in code like 
`Set.of_list (List.map_rev ...)`.

**Comment by nojb** [comment]
> > Apologies for the drive-by commenting, but in the past, I found that the only way to reason easily about all this "rev" stuff is to have a dedicated type 'a RevList.t for "reversed lists".
> 
> Yeah, i've had something like that in mind as well:
> 
> ```ocaml
> type 'a revlist =
>   | In_order of 'a list
>   | Rev of 'a list
> 
> val to_list : 'a revlist -> 'a list (* reverses if necessary *)
> val rev : 'a revlist -> 'a revlist (* flips in_order and rev *)
> val map : 'a revlist -> f:('a -> 'b) -> 'b revlist
> ..
> ```
> 

Actually what I had in mind was just:
```
module RevList : sig
  type 'a t
  val to_list : 'a t -> 'a list
  val map : ('a -> 'b) -> 'a t -> 'b list (* for example *)
  ...
end = struct
  type 'a t = 'a list (* kept in reverse order *)
  let to_list t = List.rev t
  let map f t = List.rev_map f t
  ...
end
```
so when you have an operation that naturally returns elements in reverse order you would return a `'a RevList.t`. But at the end of the day you will need to call `RevList.to_list` in order to get a "normal" list. Using a different type makes it impossible to get the `List.rev`'s mixed up.

> But it seems more trouble than its worth.

I agree that it is overkill here.

**Pull request merged**